source/opt/scalar_analysis_nodes.h

```diff
@@ -0,0 +1,284 @@
+  const_iterator cbegin() { return children_.cbegin(); }
+  const_iterator cend() { return children_.cend(); }
+
+  // Iterator to iterate over the entire DAG. Even though we are using the tree
+  // iterator it should still be safe to iterate over. However, nodes with
+  // multiple parents will be visited multiple times, unlike in a tree.
+  using dag_iterator = TreeDFIterator<SENode>;
+  using const_dag_iterator = TreeDFIterator<const SENode>;
+
+  // Iterate over all child nodes in the graph.
+  dag_iterator graph_begin() { return dag_iterator(this); }
+  dag_iterator graph_end() { return dag_iterator(); }
+  const_dag_iterator graph_begin() const { return graph_cbegin(); }
+  const_dag_iterator graph_end() const { return graph_cend(); }
+  const_dag_iterator graph_cbegin() const { return const_dag_iterator(this); }
+  const_dag_iterator graph_cend() const { return const_dag_iterator(); }
+
+  // Return the vector of immediate children.
+  const std::vector<SENode*>& GetChildren() const { return children_; }
+  std::vector<SENode*>& GetChildren() { return children_; }
+
+// Implements a casting method for each type.
+#define DeclareCastMethod(target)                  \
+  virtual target* As##target() { return nullptr; } \
+  virtual const target* As##target() const { return nullptr; }
+  DeclareCastMethod(SEConstantNode);
+  DeclareCastMethod(SERecurrentNode);
+  DeclareCastMethod(SEAddNode);
+  DeclareCastMethod(SEMultiplyNode);
+  DeclareCastMethod(SENegative);
+  DeclareCastMethod(SEValueUnknown);
+  DeclareCastMethod(SECantCompute);
+#undef DeclareCastMethod
+
+  // Get the analysis which has this node in its cache.
+  inline opt::ScalarEvolutionAnalysis* GetParentAnalysis() const {
+    return parent_analysis_;
+  }
+
+ protected:
+  std::vector<SENode*> children_;
+
+  opt::ScalarEvolutionAnalysis* parent_analysis_;
+};
+
+// Function object to handle the hashing of SENodes. Hashing algorithm hashes
+// the type (as a string), the literal value of any constants, and the child
+// pointers which are assumed to be unique.
+struct SENodeHash {
+  size_t operator()(const std::unique_ptr<SENode>& node) const;
+  size_t operator()(const SENode* node) const;
+};
+
+// A node representing a constant integer.
+class SEConstantNode : public SENode {
+ public:
+  SEConstantNode(opt::ScalarEvolutionAnalysis* parent_analysis, int64_t value)
+      : SENode(parent_analysis), literal_value_(value) {}
+
+  SENodeType GetType() const final { return Constant; }
+
+  int64_t FoldToSingleValue() const { return literal_value_; }
+
+  SEConstantNode* AsSEConstantNode() override { return this; }
+  const SEConstantNode* AsSEConstantNode() const override { return this; }
+
+ protected:
+  int64_t literal_value_;
+};
+
+// A node represeting a recurrent expression in the code. A recurrent expression
+// is an expression with a loop variant as one of its terms, such as an
+// induction variable. The actual value of a recurrent expression is coefficent_
```

Comment:
The sentence "A recurrent expression is ..." is a little confusing because it is too general.  The sentence that follows is much more concrete.  That is the part that actually told me what you meant.  It seem like what you really want is that an SERecurrentNode represents an expression whose value can be expressed as a linear expression.